frontend/sbom/sbom.go

```diff
@@ -0,0 +1,91 @@
+package sbom
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/moby/buildkit/client/llb"
+	gatewaypb "github.com/moby/buildkit/frontend/gateway/pb"
+	"github.com/moby/buildkit/solver/result"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+type Scanner[T any] func(ctx context.Context, resolver llb.ImageMetaResolver, target *result.Result[T], keys []string) error
+
+func CreateScanner[T any](scanner reference.Named, toState func(context.Context, T) (llb.State, error), fromState func(context.Context, llb.State) (T, error)) Scanner[T] {
+	if scanner == nil {
+		return nil
+	}
+
+	return func(ctx context.Context, resolver llb.ImageMetaResolver, target *result.Result[T], keys []string) error {
+		scanner = reference.TagNameOnly(scanner)
```

Comment:
I don't think this is needed in here

Reply:
It's not needed for `llb.Image`, but I think we need it for `ResolveImageConfig` - I'll move the ref parsing logic out of here though, and into the caller, so we can just pass a string for the ref in.